// @switchboard-xyz/on-demand/dist/esm/utils/signatureAuth.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import bs58 from 'bs58';
import * as crypto from 'crypto';
import nacl from 'tweetnacl';
/**
 * Manages signature-based authentication for Switchboard service
 * Signs blockhash periodically to prove subscription and freshness
 */
export class SignatureAuth {
    constructor(config) {
        this.lastRefreshTime = 0;
        this.keypair = config.keypair;
        this.connection = config.connection;
        this.refreshInterval = config.refreshInterval || 5 * 60 * 1000; // 5 minutes default (signature freshness window)
    }
    /**
     * Check if signature auth is configured and enabled
     */
    isEnabled() {
        return !!(this.keypair && this.connection);
    }
    /**
     * Get the current auth headers, refreshing if necessary
     */
    getAuthHeaders() {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this.isEnabled()) {
                return null;
            }
            // Check if we need to refresh
            const now = Date.now();
            if (!this.currentAuthData ||
                now - this.lastRefreshTime >= this.refreshInterval) {
                yield this.refresh();
            }
            if (!this.currentAuthData) {
                return null;
            }
            // Return auth headers
            return {
                'X-Switchboard-Signature': this.currentAuthData.signature,
                'X-Switchboard-Pubkey': this.currentAuthData.publicKey,
                'X-Switchboard-Blockhash': this.currentAuthData.blockhash,
                'X-Switchboard-Timestamp': this.currentAuthData.timestamp.toString(),
            };
        });
    }
    /**
     * Get the current auth data as an object (for WebSocket messages)
     */
    getAuthData() {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this.isEnabled()) {
                return null;
            }
            // Check if we need to refresh
            const now = Date.now();
            if (!this.currentAuthData ||
                now - this.lastRefreshTime >= this.refreshInterval) {
                yield this.refresh();
            }
            return this.currentAuthData || null;
        });
    }
    /**
     * Refresh the signature by fetching latest blockhash and signing
     */
    refresh() {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this.isEnabled()) {
                return;
            }
            try {
                // Fetch latest blockhash
                const { blockhash } = yield this.connection.getLatestBlockhash();
                // Create message to sign
                const timestamp = Date.now();
                const message = this.createMessage(blockhash, timestamp);
                // Sign the message with Ed25519 using tweetnacl
                const signatureBytes = nacl.sign.detached(message, this.keypair.secretKey);
                // Store the auth data (use base58 encoding - Solana standard)
                this.currentAuthData = {
                    signature: bs58.encode(signatureBytes),
                    publicKey: this.keypair.publicKey.toBase58(),
                    blockhash,
                    timestamp,
                };
                this.lastRefreshTime = Date.now();
            }
            catch (error) {
                console.error('Failed to refresh signature auth:', error);
                throw error;
            }
        });
    }
    /**
     * Create the message to sign
     * Format: SHA256(blockhash:timestamp)
     */
    createMessage(blockhash, timestamp) {
        // Message format: blockhash:timestamp
        const messageStr = `${blockhash}:${timestamp}`;
        const messageBuffer = Buffer.from(messageStr, 'utf-8');
        // Hash the message with SHA256 (matches Rust implementation)
        const hash = crypto.createHash('sha256').update(messageBuffer).digest();
        return new Uint8Array(hash);
    }
    /**
     * Start automatic refresh timer
     */
    startAutoRefresh() {
        if (!this.isEnabled()) {
            return;
        }
        // Clear any existing timer
        this.stopAutoRefresh();
        // Refresh immediately
        this.refresh().catch(err => console.error('Initial signature refresh failed:', err));
        // Set up periodic refresh
        this.refreshTimer = setInterval(() => {
            this.refresh().catch(err => console.error('Signature refresh failed:', err));
        }, this.refreshInterval);
    }
    /**
     * Stop automatic refresh timer
     */
    stopAutoRefresh() {
        if (this.refreshTimer) {
            clearInterval(this.refreshTimer);
            this.refreshTimer = undefined;
        }
    }
    /**
     * Cleanup resources
     */
    destroy() {
        this.stopAutoRefresh();
        this.currentAuthData = undefined;
    }
}
